### PR TITLE
improve error handling in cli

### PR DESCRIPTION
**Pull request by drahnr**
Introduces explicit error type variants for all `Other("$content")` types, uses `thiserror` and cleans up a slew of unused dependencies. Also adds a chainable version `Forwarded(Box<dyn Error>)` which is helpful for using this with the ability to display a error chain in arbitrary format.

Follow up PRs might want to visit the exact wording of errors again, to make sure these are descriptive enough.

### CODE REVIEW

File: client/cli/src/error.rs [hunk 1]
```diff
@@ -18,61 +18,87 @@
...
 
-
+use sp_core::crypto;
 
 /// Result type alias for the CLI.
 pub type Result<T> = std::result::Result<T, Error>;
 
 /// Error type for the CLI.
-#[derive(Debug, derive_more::Display, derive_more::From)]
+#[derive(Debug, thiserror::Error)]
```
Comment (bkchr) [inline client/cli/src/error.rs:27]: So why do we need this? :D

**Review by bkchr — COMMENTED** [review]
I don't understand why we need `thiserror` here?

File: client/cli/src/error.rs (hunk 1 above)
Comment (drahnr) [reply client/cli/src/error.rs:27]: Technically we do not _need_ it.
For consistency, for convenience, for not having to implement `std::error::Error` manually and update it manually it seemed reasonable to convert to `thiserror`.

**Review by ordian — APPROVED** [review]
LGTM

File: client/cli/src/error.rs (hunk 1 above)
Comment (drahnr) [reply client/cli/src/error.rs:27]: @bkchr is that a strong opposition or just a `I don't want random garbage in there` call?

**Review by bkchr — APPROVED** [review]

File: client/cli/src/error.rs (hunk 1 above)
Comment (bkchr) [reply client/cli/src/error.rs:27]: > For consistency, for convenience, for not having to implement std::error::Error manually and update it manually it seemed reasonable to convert to thiserror.

Thought it was doing more. So, it should be fine.

File: client/cli/src/error.rs [hunk 2]
```diff
@@ -18,61 +18,87 @@
...
 	/// Invalid listen multiaddress
-	#[display(fmt="Invalid listen multiaddress")]
-	#[from(ignore)]
+	#[error("Invalid listen multiaddress")]
 	InvalidListenMultiaddress,
-	/// Other uncategorized error.
-	#[from(ignore)]
+	/// Application specific error chain sequence forwarder.
+	#[error(transparent)]
+	Forwarded(#[from] Box<dyn std::error::Error>),
```
Comment (bkchr) [inline client/cli/src/error.rs:52]:
```suggestion
	Application(#[from] Box<dyn std::error::Error>),
```

This or something else? `Forwarded` sounds weird.

**Comment by bkchr** [comment]
We should find a better name for `Forwarded`, otherwise it looks okay.

File: client/cli/src/error.rs (hunk 1 above)
Comment (drahnr) [reply client/cli/src/error.rs:27]: That's the beauty, it literally removes the chores.

**Comment by drahnr** [comment]
bot merge

**Pull request merged**